src/python/pants/build_graph/build_graph.py

```diff
@@ -479,11 +545,11 @@ def transitive_subgraph_of_addresses_bfs(self, addresses, predicate=None, levele
         continue
       if walk.do_work_once(address):
         ordered_closure.add(target)
-      for addr in self._target_dependencies_by_address[address]:
-        if walk.expanded_or_worked(addr):
+      for dep_address in self._target_dependencies_by_address[address]:
+        if walk.expanded_or_worked(dep_address):
           continue
-        if not leveled_predicate or leveled_predicate(self._target_by_address[addr], level):
-          to_walk.append((level + 1, addr))
+        if walk.dep_predicate(target, self._target_by_address[dep_address], level):
```

Comment:
I expect that a fair amount of your slowdown is switching from `if not predicate` to always calling the method... unfortunately, the null-object pattern is a bit more expensive in python than elsewhere =(

I would rather land this and then do holistic optimization that micro-optimize here too much though.

Reply:
Yeah. My goal here is to make the predicate interactions super clear, and then we can denormalize from there as necessary.